src/openms/source/CHEMISTRY/EnzymesDB.cpp

```diff
@@ -0,0 +1,331 @@
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Xiao Liang  $
+// $Authors: Xiao Liang $
+// --------------------------------------------------------------------------
+//
+
+#include <OpenMS/CHEMISTRY/EnzymesDB.h>
+#include <OpenMS/CHEMISTRY/Enzyme.h>
+#include <OpenMS/SYSTEM/File.h>
+#include <OpenMS/CONCEPT/LogStream.h>
+
+#include <OpenMS/DATASTRUCTURES/Param.h>
+#include <OpenMS/DATASTRUCTURES/ListUtils.h>
+#include <OpenMS/FORMAT/ParamXMLFile.h>
+
+
+#include <vector>
+#include <fstream>
+
+using namespace std;
+
+namespace OpenMS
+{
+  EnzymesDB::EnzymesDB()
+  {
+    readEnzymesFromFile_("CHEMISTRY/Enzymes.xml");
+    buildEnzymeNames_();
+  }
+
+  EnzymesDB::~EnzymesDB()
+  {
+    clear_();
+  }
+
+  const Enzyme* EnzymesDB::getEnzyme(const String& name) const
+  {
+    if (enzyme_names_.find(name) == enzyme_names_.end())
+    {
+      throw Exception::ElementNotFound(__FILE__, __LINE__, __PRETTY_FUNCTION__, "Enzyme name cannot be found. '");
+    }
+    return enzyme_names_.at(name);
+  }
+
+  const Enzyme* EnzymesDB::getEnzymeByRegEx(const String & cleavage_regex) const
+  {
+    if (!enzyme_regex_.has(cleavage_regex))
+    {
+      throw Exception::IllegalArgument(__FILE__, __LINE__, __PRETTY_FUNCTION__, String("Enzyme with regex "
+                                                                                       + cleavage_regex + " was not registered in Enzyme DB, register first!").c_str());
+    }
+    return enzyme_regex_[cleavage_regex];
+  }
+
+  void EnzymesDB::setEnzymes(const String& file_name)
+  {
+    clearEnzymes_();
+    readEnzymesFromFile_(file_name);
+    buildEnzymeNames_();
+  }
+
+  void EnzymesDB::addEnzyme(const Enzyme& enzyme)
+  {
+    Enzyme* r = new Enzyme(enzyme);
+    addEnzyme_(r);
```

Comment:
why copy the enzyme here? why not

```addEnzyme_(&enzyme); ```

that should work as well ? are you changing the enzyme? see also Chris' comment below